src/librustc_mir/transform/promote_consts.rs

```diff
@@ -118,8 +122,8 @@ impl Candidate {
     /// Returns `true` if we should use the "explicit" rules for promotability for this `Candidate`.
     fn forces_explicit_promotion(&self) -> bool {
```

Comment:
As discussed before, could you add a `delay_span_bug` somewhere that ensure that when explicit promotion fails, compilation also fails?

Reply:
You mean https://github.com/rust-lang/rust/blob/bd18bc9a4c142100a7230e1fd1cfe6be08214693/src/librustc_codegen_ssa/mir/operand.rs#L447-L455 :rofl: ?

Reply:
No I don't think so... I mean what we discussed [recently here](https://github.com/rust-lang/rust/pull/70042#issuecomment-600620211).

Reply:
More seriously. We'd need to match on the `DefKind` and probably somehow encode the difference between explicit and implicit promotion: https://github.com/rust-lang/rust/blob/master/src/librustc_mir/const_eval/eval_queries.rs#L354-L357

Reply:
Actually, we could just carry this information in the `mir::Body`,

Reply:
What I would have thought we could do is to change this code here

https://github.com/rust-lang/rust/blob/a4e551ef26e9ba0c2f536e67e6bd33447d8a148a/src/librustc_mir/transform/promote_consts.rs#L743

and add something like
```rust
if validator.explicit && !is_promotable { delay_span_bug(); }
```

Reply:
By "promotion fails" I did not mean "evaluating the promoted fails", I meant "promotion was found not to be possible". We have the key invariant that explicit promotion rules may only be used if, when these rules fail, the program would anyway fail to compile.

Reply:
Good catch!

```
-> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:77:5
   |
LL | /     fn no_impl_copy_empty_value_no_elements() {
LL | |         let x = None;
LL | |         let arr: [Option<Bar>; 0] = [x; 0];
LL | |     }
   | |_____^

error: internal compiler error: Explicit promotion requested, but failed to promote
  --> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:82:5
   |
LL | /     fn no_impl_copy_empty_value_single_element() {
LL | |         let x = None;
LL | |         let arr: [Option<Bar>; 1] = [x; 1];
LL | |     }
   | |_____^

error: internal compiler error: Explicit promotion requested, but failed to promote
  --> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:87:5
   |
LL | /     fn no_impl_copy_value_no_elements() {
LL | |         let x = Some(Bar);
LL | |         let arr: [Option<Bar>; 0] = [x; 0];
LL | |     }
   | |_____^

error: internal compiler error: Explicit promotion requested, but failed to promote
  --> /home/oli/rustc/rust7/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs:92:5
   |
LL | /     fn no_impl_copy_value_single_element() {
LL | |         let x = Some(Bar);
LL | |         let arr: [Option<Bar>; 1] = [x; 1];
LL | |     }
   | |_____^
```